tss-esapi/src/abstraction/public.rs

```diff
@@ -2,127 +2,250 @@
+                unique, parameters, ..
+            } => {
+                let exponent = match parameters.exponent() {
+                    RsaExponent::ZERO_EXPONENT => BigUint::from(RSA_DEFAULT_EXP),
+                    _ => BigUint::from(parameters.exponent().value()),
+                };
+                let modulus = BigUint::from_bytes_be(unique.as_bytes());
+
+                let public_key = RsaPublicKey::new(modulus, exponent)
+                    .map_err(|_| Error::local_error(WrapperErrorKind::InvalidParam))?;
+
+                Ok(public_key)
+            }
+            _ => Err(Error::local_error(WrapperErrorKind::UnsupportedParam)),
+        }
     }
 }
 
-impl TryFrom<Public> for SubjectPublicKeyInfo {
+impl TryFrom<&Public> for SubjectPublicKeyInfoOwned {
     type Error = Error;
 
-    /// Converts [`crate::structures::Public::Rsa`] and [`crate::structures::Public::Ecc`] to [`picky_asn1_x509::SubjectPublicKeyInfo`].
+    /// Converts [`crate::structures::Public::Rsa`] and [`crate::structures::Public::Ecc`] to [`x509_cert::spki::SubjectPublicKeyInfoOwned`].
     ///
     /// # Details
-    /// The result can be used to convert TPM public keys to DER using `picky_asn1_der`.
+    /// The result can be used to convert TPM public keys to DER using `x509-cert`.
     ///
     /// # Errors
     /// * if other instances of [`crate::structures::Public`] are used `UnsupportedParam` will be returned.
-    fn try_from(value: Public) -> Result<Self, Self::Error> {
-        let decoded_key = public_to_decoded_key(&value)?;
-
-        match (value, decoded_key) {
-            (Public::Rsa { .. }, DecodedKey::RsaPublicKey(key)) => Ok(SubjectPublicKeyInfo {
-                algorithm: AlgorithmIdentifier::new_rsa_encryption(),
-                subject_public_key: PublicKey::Rsa(key.into()),
-            }),
-            (Public::Ecc { parameters, .. }, DecodedKey::EcPoint(point)) => {
-                Ok(SubjectPublicKeyInfo {
-                    algorithm: AlgorithmIdentifier::new_elliptic_curve(EcParameters::NamedCurve(
-                        curve_oid(parameters.ecc_curve())?.into(),
-                    )),
-                    subject_public_key: PublicKey::Ec(BitString::with_bytes(point).into()),
-                })
+    fn try_from(value: &Public) -> Result<Self, Self::Error> {
+        match value {
+            #[cfg(feature = "rsa")]
+            Public::Rsa { .. } => {
+                let public_key = RsaPublicKey::try_from(value)?;
+
+                Ok(public_key
+                    .to_public_key_der()
+                    .map_err(|_| Error::local_error(WrapperErrorKind::InvalidParam))?
+                    .decode_msg::<Self>()
+                    .map_err(|_| Error::local_error(WrapperErrorKind::InvalidParam))?)
+            }
+            #[allow(unused)]
+            Public::Ecc { parameters, .. } => {
+                macro_rules! read_key {
+                    ($curve:expr, $key_type:ty) => {
```

Comment:
Why do we need `$curve` if it's not used anywhere? (also the `allow(unused)` lint)

Reply:
The `allow(unused)` is if you compile with `features = ["rustcrypto", "rsa"]` and opt-out of ECC, then the `read_key` macro gets unused.